File: CS/Dashboard_FiscalFunctions/Fiscal Functions/GetFiscalQuarterFunction.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Dashboard_FiscalFunctions {
    class GetFiscalQuarterFunction : GetFiscalDateFunction {
        public GetFiscalQuarterFunction(int startDay, int startMonth) : base(startDay, startMonth) {}

        public override string Name => "GetFiscalQuarter";

        public override string Description => "GetFiscalQuarter(DateTime) \n Extracts a fiscal quarter from the defined DateTime.";

        protected override int GetFiscal(DateTime current) {
            int currentMonth = current.Month;
            if ((current.Month + 12 - startMonth) % 3 == 0) {
                currentMonth += current.Day < startDay ? -1 : 1;
            }
            int offset = currentMonth < startMonth ? 12 : 0;
            return (currentMonth - startMonth + offset) / 3 + 1;
        }
    }
}

File: VB/Dashboard_FiscalFunctions/Fiscal Functions/GetFiscalQuarterFunction.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks

Namespace Dashboard_FiscalFunctions
	Friend Class GetFiscalQuarterFunction
		Inherits GetFiscalDateFunction

		Public Sub New(ByVal startDay As Integer, ByVal startMonth As Integer)
			MyBase.New(startDay, startMonth)
		End Sub

		Public Overrides ReadOnly Property Name() As String
			Get
				Return "GetFiscalQuarter"
			End Get
		End Property

		Public Overrides ReadOnly Property Description() As String
			Get
				Return "GetFiscalQuarter(DateTime) " & vbLf & " Extracts a fiscal quarter from the defined DateTime."
			End Get
		End Property

		Protected Overrides Function GetFiscal(ByVal current As Date) As Integer
			Dim currentMonth As Integer = current.Month
			If (current.Month + 12 - startMonth) Mod 3 = 0 Then
				currentMonth += If(current.Day < startDay, -1, 1)
			End If
			Dim offset As Integer = If(currentMonth < startMonth, 12, 0)
			Return (currentMonth - startMonth + offset) \ 3 + 1
		End Function
	End Class
End Namespace
